### PR TITLE
[#157242214] stop crawler repositories end

### PR DESCRIPTION
**Pull request by r3vit**
Stop crawler when repositories end.
Add goroutine that performs a continuous check on waiting group and closes the channel when it's counter is zero.

### CODE REVIEW

File: crawler/clientApi.go [hunk 1]
```diff
@@ -8,7 +8,7 @@ import (
 	"path/filepath"
 	"plugin"
 
-	log "github.com/sirupsen/logrus"
+	"github.com/prometheus/common/log"
```
Comment (lussoluca) [inline crawler/clientApi.go]: uhm, why?

File: crawler/crawler.go [hunk 2]
```diff
@@ -109,6 +104,9 @@ func checkAvailability(repository Repository, processedCounter prometheus.Counte
 			log.Warn("Validator errors:" + err.Error())
 		}
 	}
+
+	// Defer waiting group close.
+	wg.Done()
```
Comment (lussoluca) [inline crawler/crawler.go]: maybe a `defer wg.Done()` at the beginning of the function is more safe

File: cmd/single.go [hunk 3]
```diff
@@ -37,15 +38,21 @@ Beware! May take days to complete.`,
...
 		// Initiate a channel of repositories.
 		repositories := make(chan crawler.Repository, 100)
+		// Prepare WaitGroup.
+		var wg sync.WaitGroup
 
 		// Process each domain service.
 		for _, domain := range domains {
 			if domain.Id == domainID {
-				go crawler.ProcessDomain(domain, repositories)
+				wg.Add(1)
```
Comment (lussoluca) [inline cmd/single.go]: why we need to count also the domains in the WaitGroup?

File: cmd/single.go [hunk 4]
```diff
@@ -37,15 +38,21 @@ Beware! May take days to complete.`,
...
+				go crawler.ProcessDomain(domain, repositories, &wg)
 			}
 		}
 
+		// Goroutine that constatly check if all the processes are terminated.
+		go crawler.WaitingLoop(repositories, &wg)
+
 		// Process the repositories in order to retrieve publiccode.yml.
-		crawler.ProcessRepositories(repositories)
+		crawler.ProcessRepositories(repositories, &wg)
```
Comment (lussoluca) [inline cmd/single.go]: can we start `crawler.ProcessRepositories` in a goroutine and move here the code that wait and close the channel?

File: cmd/single.go (hunk 3 above)
Comment (r3vit) [reply cmd/single.go]: Special case: after filling the repositories channel with repos all the domains stop to reply to our requests. All the single files are checked and the WG is empty. End.
This wg.Add() will perform an additional check on the domain level. Does not stop until the end is reached (and retry continuously until a domain ends.)

File: crawler/crawler.go (hunk 2 above)
Comment (r3vit) [reply crawler/crawler.go]: defer is called after the surrounding function returns (https://blog.golang.org/defer-panic-and-recover). As goroutine there is no return in function (only channel communication). The defer will never be called.

File: crawler/clientApi.go (hunk 1 above)
Comment (r3vit) [reply crawler/clientApi.go]: Wrong auto-import. Fixed.

File: crawler/crawler.go (hunk 2 above)
Comment (lussoluca) [reply crawler/crawler.go]: yes, correct

File: cmd/single.go (hunk 4 above)
Comment (r3vit) [reply cmd/single.go]: Ok. Moved crawler.ProcessRepositories in a goroutine.
The code of WaitingLoop should be extracted from the function?

**Pull request merged**